File: backend/app/services/portfolio.py
from decimal import Decimal

from sqlmodel import Session, select

from app.models import AssetType, Investment, Transaction, TransactionType, User
from app.schemas.dashboard import AssetTypeSummary, DashboardSummary
from app.schemas.investments import InvestmentRead

MONEY_PLACES = Decimal("0.0001")
QUANTITY_PLACES = Decimal("0.000001")


def quantize_money(value: Decimal) -> Decimal:
    return value.quantize(MONEY_PLACES)


def quantize_quantity(value: Decimal) -> Decimal:
    return value.quantize(QUANTITY_PLACES)


def build_investment_read(
    investment: Investment,
    session: Session,
) -> InvestmentRead:
    transactions = session.exec(
        select(Transaction).where(Transaction.investment_id == investment.id),
    ).all()

    total_buy_quantity = sum(
        (
            transaction.quantity
            for transaction in transactions
            if transaction.transaction_type == TransactionType.BUY
        ),
        Decimal("0"),
    )
    total_sell_quantity = sum(
        (
            transaction.quantity
            for transaction in transactions
            if transaction.transaction_type == TransactionType.SELL
        ),
        Decimal("0"),
    )
    total_buy_cost = sum(
        (
            transaction.quantity * transaction.price
            for transaction in transactions
            if transaction.transaction_type == TransactionType.BUY
        ),
        Decimal("0"),
    )

    current_quantity = total_buy_quantity - total_sell_quantity
    average_buy_price = (
        total_buy_cost / total_buy_quantity
        if total_buy_quantity > 0
        else Decimal("0")
    )
    estimated_cost_basis = current_quantity * average_buy_price
    current_value = current_quantity * investment.current_price
    gain_loss = current_value - estimated_cost_basis
    performance_percentage = (
        gain_loss / estimated_cost_basis * Decimal("100")
        if estimated_cost_basis > 0
        else Decimal("0")
    )

    return InvestmentRead(
        id=investment.id,
        name=investment.name,
        symbol=investment.symbol,
        asset_type=investment.asset_type,
        current_price=quantize_money(investment.current_price),
        current_quantity=quantize_quantity(current_quantity),
        average_buy_price=quantize_money(average_buy_price),
        estimated_cost_basis=quantize_money(estimated_cost_basis),
        current_value=quantize_money(current_value),
        gain_loss=quantize_money(gain_loss),
        performance_percentage=quantize_money(performance_percentage),
    )


def build_dashboard_summary(
    *,
    current_user: User,
    session: Session,
) -> DashboardSummary:
    investments = session.exec(
        select(Investment)
        .where(Investment.user_id == current_user.id)
        .order_by(Investment.asset_type, Investment.symbol),
    ).all()
    investment_reads = [
        build_investment_read(investment=investment, session=session)
        for investment in investments
    ]

    total_current_value = sum(
        (investment.current_value for investment in investment_reads),
        Decimal("0"),
    )
    total_cost_basis = sum(
        (investment.estimated_cost_basis for investment in investment_reads),
        Decimal("0"),
    )
    total_gain_loss = total_current_value - total_cost_basis
    total_performance_percentage = (
        total_gain_loss / total_cost_basis * Decimal("100")
        if total_cost_basis > 0
        else Decimal("0")
    )

    asset_type_totals: dict[AssetType, dict[str, Decimal]] = {}
    for investment in investment_reads:
        totals = asset_type_totals.setdefault(
            investment.asset_type,
            {
                "current_value": Decimal("0"),
                "cost_basis": Decimal("0"),
                "gain_loss": Decimal("0"),
            },
        )
        totals["current_value"] += investment.current_value
        totals["cost_basis"] += investment.estimated_cost_basis
        totals["gain_loss"] += investment.gain_loss

    return DashboardSummary(
        total_current_value=quantize_money(total_current_value),
        total_cost_basis=quantize_money(total_cost_basis),
        total_gain_loss=quantize_money(total_gain_loss),
        total_performance_percentage=quantize_money(total_performance_percentage),
        asset_type_summary=[
            AssetTypeSummary(
                asset_type=asset_type,
                current_value=quantize_money(totals["current_value"]),
                cost_basis=quantize_money(totals["cost_basis"]),
                gain_loss=quantize_money(totals["gain_loss"]),
            )
            for asset_type, totals in sorted(
                asset_type_totals.items(),
                key=lambda item: item[0].value,
            )
        ],
    )

File: frontend/src/api/portfolio.ts
import { apiClient } from './client'

export type AssetType = 'STOCK' | 'BOND' | 'MUTUAL_FUND'
export type TransactionType = 'BUY' | 'SELL'
export type TrendPeriod = '1D' | '1M' | 'YTD' | 'ALL'

export type DashboardSummary = {
  total_current_value: string
  total_cost_basis: string
  total_gain_loss: string
  total_performance_percentage: string
  asset_type_summary: AssetTypeSummary[]
}

export type DashboardTrendPoint = {
  date: string
  value: string
  cost_basis: string
}

export type DashboardTrend = {
  period: TrendPeriod
  points: DashboardTrendPoint[]
}

export type AssetTypeSummary = {
  asset_type: AssetType
  current_value: string
  cost_basis: string
  gain_loss: string
}

export type Investment = {
  id: string
  name: string
  symbol: string
  asset_type: AssetType
  current_price: string
  current_quantity: string
  average_buy_price: string
  estimated_cost_basis: string
  current_value: string
  gain_loss: string
  performance_percentage: string
}

export type InvestmentCreatePayload = {
  name: string
  symbol: string
  asset_type: AssetType
  current_price: string
  initial_quantity: string
  initial_purchase_price: string
  transaction_date: string
}

export type InvestmentUpdatePayload = {
  name: string
  symbol: string
  asset_type: AssetType
  current_price: string
}

export type Transaction = {
  id: string
  investment_id: string
  investment_symbol: string
  transaction_type: TransactionType
  quantity: string
  price: string
  transaction_date: string
}

export type TransactionCreatePayload = {
  investment_id: string
  transaction_type: TransactionType
  quantity: string
  price: string
  transaction_date: string
}

export type TransactionUpdatePayload = {
  transaction_type: TransactionType
  quantity: string
  price: string
  transaction_date: string
}

export async function getDashboardSummary() {
  const response = await apiClient.get<DashboardSummary>('/dashboard/summary')
  return response.data
}

export async function getDashboardTrend(period: TrendPeriod) {
  const response = await apiClient.get<DashboardTrend>('/dashboard/trend', {
    params: { period },
  })
  return response.data
}

export async function listInvestments() {
  const response = await apiClient.get<Investment[]>('/investments')
  return response.data
}

export async function createInvestment(payload: InvestmentCreatePayload) {
  const response = await apiClient.post<Investment>('/investments', payload)
  return response.data
}

export async function updateInvestment(
  investmentId: string,
  payload: InvestmentUpdatePayload,
) {
  const response = await apiClient.put<Investment>(
    `/investments/${investmentId}`,
    payload,
  )
  return response.data
}

export async function deleteInvestment(investmentId: string) {
  await apiClient.delete(`/investments/${investmentId}`)
}

export async function listTransactions() {
  const response = await apiClient.get<Transaction[]>('/transactions')
  return response.data
}

export async function createTransaction(payload: TransactionCreatePayload) {
  const response = await apiClient.post<Transaction>('/transactions', payload)
  return response.data
}

export async function updateTransaction(
  transactionId: string,
  payload: TransactionUpdatePayload,
) {
  const response = await apiClient.put<Transaction>(
    `/transactions/${transactionId}`,
    payload,
  )
  return response.data
}

export async function deleteTransaction(transactionId: string) {
  await apiClient.delete(`/transactions/${transactionId}`)
}
